tensorflow_model_optimization/python/core/clustering/keras/cluster_integration_test.py

```diff
@@ -30,55 +33,106 @@
+            dtype="float32",
+        )
+
+    def dataset_generator(self):
+        for x, y in zip(self.x_train, self.y_train):
+            yield np.array([x]), np.array([y])
+
+    @staticmethod
+    def _verify_tflite(tflite_file, x_test):
+        interpreter = tf.lite.Interpreter(model_path=tflite_file)
+        interpreter.allocate_tensors()
+        input_index = interpreter.get_input_details()[0]["index"]
+        output_index = interpreter.get_output_details()[0]["index"]
+        x = x_test[0]
+        x = x.reshape((1,) + x.shape)
+        interpreter.set_tensor(input_index, x)
+        interpreter.invoke()
+        interpreter.get_tensor(output_index)
+
+    @keras_parameterized.run_all_keras_modes
+    def testValuesRemainClusteredAfterTraining(self):
+
+        """
+        Verifies that training a clustered model does not destroy the clusters.
+        """
+        original_model = keras.Sequential(
+            [layers.Dense(2, input_shape=(2,)), layers.Dense(2),]
+        )
+
+        clustered_model = cluster.cluster_weights(original_model, **self.params)
+
+        clustered_model.compile(
+            loss=keras.losses.categorical_crossentropy,
+            optimizer="adam",
+            metrics=["accuracy"],
+        )
+
+        clustered_model.fit(x=self.dataset_generator(), steps_per_epoch=1)
+        stripped_model = cluster.strip_clustering(clustered_model)
+        weights_as_list = stripped_model.get_weights()[0].reshape(-1,).tolist()
+        unique_weights = set(weights_as_list)
+        self.assertLessEqual(len(unique_weights), self.params["number_of_clusters"])
+
+    @keras_parameterized.run_all_keras_modes(always_skip_v1=True)
+    def testEndToEnd(self):
+
+        """
+        Test End to End clustering.
+        """
+        original_model = keras.Sequential(
+            [layers.Dense(2, input_shape=(2,)), layers.Dense(2),]
+        )
+
+        clustered_model = cluster.cluster_weights(original_model, **self.params)
+
+        clustered_model.compile(
+            loss=keras.losses.categorical_crossentropy,
+            optimizer="adam",
+            metrics=["accuracy"],
+        )
+
+        clustered_model.fit(x=self.dataset_generator(), steps_per_epoch=1)
+        stripped_model = cluster.strip_clustering(clustered_model)
+
+        _, tflite_file = tempfile.mkstemp(".tflite")
+        _, keras_file = tempfile.mkstemp(".h5")
+
+        if not compat.is_v1_apis():
+            converter = tf.lite.TFLiteConverter.from_keras_model(stripped_model)
+        else:
+            tf.keras.models.save_model(stripped_model, keras_file)
+            converter = tf.lite.TFLiteConverter.from_keras_model_file(keras_file)
+
+        converter.experimental_new_converter = True
+        tflite_model = converter.convert()
+        with open(tflite_file, "wb") as f:
+            f.write(tflite_model)
+
+        self._verify_tflite(tflite_file, self.x_train)
+
+        os.remove(keras_file)
```

Comment:
Feel free to cleanup in a followup: `os.remove` should not be necessary since tempfile will automatically delete it.